frigate/embeddings/maintainer.py

```diff
@@ -0,0 +1,180 @@
+
+                    # Embed the thumbnail
+                    self._embed_thumbnail(event_id, thumbnail, metadata)
+
+                    if (
+                        camera_config.genai.enabled
+                        and self.genai_client is not None
+                        and event.data.get("description") is None
+                    ):
+                        # Generate the description. Call happens in a thread since it is network bound.
+                        threading.Thread(
+                            target=self._embed_description,
+                            name=f"_embed_description_{event.id}",
+                            daemon=True,
+                            args=(
+                                event,
+                                [
+                                    data["thumbnail"]
+                                    for data in self.tracked_events.get(
+                                        event_id, [{"thumbnail": thumbnail}]
+                                    )
+                                ],
+                                metadata,
+                            ),
+                        ).start()
+
+                # Delete tracked events based on the event_id
+                if event_id in self.tracked_events:
+                    del self.tracked_events[event_id]
+
+    def _create_thumbnail(self, yuv_frame, box, height=500) -> Optional[bytes]:
+        """Return jpg thumbnail of a region of the frame."""
+        frame = cv2.cvtColor(yuv_frame, cv2.COLOR_YUV2BGR_I420)
+        region = calculate_region(
+            frame.shape, box[0], box[1], box[2], box[3], height, multiplier=1.4
+        )
+        frame = frame[region[1] : region[3], region[0] : region[2]]
+        width = int(height * frame.shape[1] / frame.shape[0])
+        frame = cv2.resize(frame, dsize=(width, height), interpolation=cv2.INTER_AREA)
+        ret, jpg = cv2.imencode(".jpg", frame, [int(cv2.IMWRITE_JPEG_QUALITY), 100])
+
+        if ret:
+            return jpg.tobytes()
+
+        return None
+
+    def _embed_thumbnail(self, event_id: str, thumbnail: bytes, metadata: dict) -> None:
+        """Embed the thumbnail for an event."""
+
+        # Encode the thumbnail
+        img = np.array(Image.open(io.BytesIO(thumbnail)).convert("RGB"))
+        self.embeddings.thumbnail.upsert(
+            images=[img],
+            metadatas=[metadata],
+            ids=[event_id],
+        )
+
+    def _embed_description(
+        self, event: Event, thumbnails: list[bytes], metadata: dict
+    ) -> None:
+        """Embed the description for an event."""
+
+        description = self.genai_client.generate_description(thumbnails, metadata)
+
+        if description is None:
+            logger.debug("Failed to generate description for %s", event.id)
+            return
+
+        # Update the event to add the description
+        event.data["description"] = description
+        event.save()
```

Comment:
Yeah. I've verified that it seems to work that way. I am doing this because we are running the API calls in a thread and there is no return value. I could publish this to the inter process communicator and have it do the writes but it felt like a lot of extra code. Happy to make changes here if you want though.

Reply:
the problem that we ran in to was that the main process handles all DB operations using a queue which often led to the database connection on the secondary process failing due to it being locked, so we moved to everything being done through that main process and the communicator.